data-model/src/main/java/io/micronaut/data/annotation/QueryResult.java

```diff
@@ -45,6 +47,11 @@
      */
     String column() default DEFAULT_COLUMN;
 
+    /**
+     * @return The data type that resulting column will hold. It helps to pick proper column reader based on result data type
+     */
+    DataType dataType() default DataType.JSON;
```

Comment:
I think this should alias to `@TypeDef`, which should be a default way to provide the type of the result:

`@TypeDef(dataType = INTEGER ARRAY) @Query("select array ..") Integer[] myArray`